### PR TITLE
Prevent duplicate variables in Ansible Playbooks

### PR DESCRIPTION
**Pull request by jan-cerny**
The same XCCDF value can appear in multiple rules, which means
there can be same variable in multiple Ansible task. But when
we generate a Playbook we should put the variable only once.
Ansible produces a warning if there are multiple definitions
of the same variable.

See https://github.com/OpenSCAP/openscap/pull/896 for further context if you are interested.

### CODE REVIEW

File: src/XCCDF_POLICY/xccdf_policy_remediate.c [hunk 1]
```diff
@@ -733,37 +710,55 @@ static inline int _xccdf_policy_rule_generate_fix(struct xccdf_policy *policy, s
...
 				xccdf_rule_get_id(rule));
 		xccdf_fix_free(cfix);
 		return 1;
 	}
 	xccdf_fix_free(cfix);
+	return 0;
+}
 
-	int ret = _write_fix_header_to_fd(template, output_fd, rule, current, total);
+static int _xccdf_policy_rule_generate_fix(struct xccdf_policy *policy, struct xccdf_rule *rule, const char *template, int output_fd, unsigned int current, unsigned int total, bool ansible_variable_mode)
```
Comment (jan-cerny) [inline src/XCCDF_POLICY/xccdf_policy_remediate.c]: ansible_variable_mode is unused in this function, please remove it

File: src/XCCDF_POLICY/xccdf_policy_remediate.c [hunk 2]
```diff
@@ -977,30 +972,45 @@ int xccdf_policy_generate_fix(struct xccdf_policy *policy, struct xccdf_result *
 
 	// In Ansible we have to generate variables first and then tasks
 	if (strcmp(sys, "urn:xccdf:fix:script:ansible") == 0) {
+		struct oscap_list *variables = oscap_list_new();
+		struct oscap_list *tasks = oscap_list_new();
+		struct oscap_iterator *rules_to_fix_it = oscap_iterator_new(rules_to_fix);
```
Comment (matejak) [inline src/XCCDF_POLICY/xccdf_policy_remediate.c]: I propose to do some function extraction here. What about introducing `get_vars_and_tasks_from_rules(rules_to_fix, variables, tasks)` here?

**Comment by jan-cerny** [comment]
@matejak I have tried to do a function refactoring.

File: src/XCCDF_POLICY/xccdf_policy_remediate.c [hunk 3]
```diff
@@ -733,37 +710,55 @@ static inline int _xccdf_policy_rule_generate_fix(struct xccdf_policy *policy, s
...
 				xccdf_rule_get_id(rule));
 		xccdf_fix_free(cfix);
 		return 1;
 	}
 	xccdf_fix_free(cfix);
+	return 0;
+}
 
-	int ret = _write_fix_header_to_fd(template, output_fd, rule, current, total);
+static int _xccdf_policy_rule_generate_fix(struct xccdf_policy *policy, struct xccdf_rule *rule, const char *template, int output_fd, unsigned int current, unsigned int total)
```
[the remark this answers is not in the document]
Comment (jan-cerny) [reply src/XCCDF_POLICY/xccdf_policy_remediate.c]: Nice, I'll look into that.

**Comment by matejak** [comment]
@jan-cerny And please don't forget to rebase, tests should pass now :-)

**Comment by matejak** [comment]
Looks great, thank you!

**Pull request merged**